pkg/pbflow/proto.go

```diff
@@ -0,0 +1,202 @@
+		pbflowRecord.Network.DstAddr = &IP{IpFamily: &IP_Ipv6{Ipv6: fr.Id.DstIp[:]}}
+	} else {
+		pbflowRecord.Network.SrcAddr = &IP{IpFamily: &IP_Ipv4{Ipv4: flow.IntEncodeV4(fr.Id.SrcIp)}}
+		pbflowRecord.Network.DstAddr = &IP{IpFamily: &IP_Ipv4{Ipv4: flow.IntEncodeV4(fr.Id.DstIp)}}
+	}
+	return &pbflowRecord
+}
+
+func PBToFlow(pb *Record) *flow.Record {
+	if pb == nil {
+		return nil
+	}
+	out := flow.Record{
+		RawRecord: flow.RawRecord{
+			Id: ebpf.BpfFlowId{
+				Direction:         uint8(pb.Direction),
+				EthProtocol:       uint16(pb.EthProtocol),
+				TransportProtocol: uint8(pb.Transport.Protocol),
+				SrcMac:            macToUint8(pb.DataLink.GetSrcMac()),
+				DstMac:            macToUint8(pb.DataLink.GetDstMac()),
+				SrcIp:             ipToIPAddr(pb.Network.GetSrcAddr()),
+				DstIp:             ipToIPAddr(pb.Network.GetDstAddr()),
+				SrcPort:           uint16(pb.Transport.SrcPort),
+				DstPort:           uint16(pb.Transport.DstPort),
+				IcmpType:          uint8(pb.IcmpType),
+				IcmpCode:          uint8(pb.IcmpCode),
+			},
+			Metrics: ebpf.BpfFlowMetrics{
+				Bytes:   pb.Bytes,
+				Packets: uint32(pb.Packets),
+				Flags:   uint16(pb.Flags),
+				Dscp:    uint8(pb.Network.Dscp),
+				PktDrops: ebpf.BpfPktDropsT{
+					Bytes:           pb.PktDropBytes,
+					Packets:         uint32(pb.PktDropPackets),
+					LatestFlags:     uint16(pb.PktDropLatestFlags),
+					LatestState:     uint8(pb.PktDropLatestState),
+					LatestDropCause: pb.PktDropLatestDropCause,
+				},
+				DnsRecord: ebpf.BpfDnsRecordT{
+					Id:      uint16(pb.DnsId),
+					Flags:   uint16(pb.DnsFlags),
+					Errno:   uint8(pb.DnsErrno),
+					Latency: uint64(pb.DnsLatency.AsDuration()),
+				},
+			},
+		},
+		TimeFlowStart: pb.TimeFlowStart.AsTime(),
+		TimeFlowEnd:   pb.TimeFlowEnd.AsTime(),
+		AgentIP:       pbIPToNetIP(pb.AgentIp),
+		Duplicate:     pb.Duplicate,
+		Interface:     pb.Interface,
+		TimeFlowRtt:   pb.TimeFlowRtt.AsDuration(),
+		DNSLatency:    pb.DnsLatency.AsDuration(),
+	}
+
+	if len(pb.GetDupList()) != 0 {
+		for _, entry := range pb.GetDupList() {
+			intf := entry.Interface
+			dir := uint8(entry.Direction)
+			out.DupList = append(out.DupList, map[string]uint8{intf: dir})
+		}
+	}
+	return &out
+}
+
+// Mac bytes are encoded in the same order as in the array. This is, a Mac
+// like 11:22:33:44:55:66 will be encoded as 0x112233445566
+func macToUint64(m *[flow.MacLen]uint8) uint64 {
+	return uint64(m[5]) |
+		(uint64(m[4]) << 8) |
+		(uint64(m[3]) << 16) |
+		(uint64(m[2]) << 24) |
+		(uint64(m[1]) << 32) |
+		(uint64(m[0]) << 40)
+}
+
+func agentIP(nip net.IP) *IP {
+	if ip := nip.To4(); ip != nil {
+		return &IP{IpFamily: &IP_Ipv4{Ipv4: binary.BigEndian.Uint32(ip)}}
+	}
+	// IPv6 address
+	return &IP{IpFamily: &IP_Ipv6{Ipv6: nip}}
+}
+
+func pbIPToNetIP(ip *IP) net.IP {
+	if ip.GetIpv6() != nil {
+		return net.IP(ip.GetIpv6())
+	}
+	n := ip.GetIpv4()
+	return net.IPv4(
```

Comment:
is it any cleaner like this
```go
byteArray := make([]byte, 4)
binary.NativeEndian.PutUint32(byteArray, n)
```
or is it possible to maye it byte type in pb like ipv6 so we can avoid all of this ?

Reply:
I can do your first suggestion (I'm not sure about the other, if the way it is was initially done for good reasons or not..)

But I guess I should either use LittleEndian or BigEndian but not Native, as that would make it arch-dependent (and thus fail on BE archs)